presto-main/src/main/java/io/prestosql/sql/planner/QueryPlanner.java

```diff
@@ -804,10 +804,17 @@ private PlanBuilder window(PlanBuilder subPlan, List<FunctionCall> windowFunctio
 
             List<Symbol> sourceSymbols = subPlan.getRoot().getOutputSymbols();
             ImmutableList.Builder<Symbol> orderBySymbols = ImmutableList.builder();
-            orderBySymbols.addAll(orderings.keySet());
+            orderBySymbols.addAll(sortOrders.keySet());
             Optional<OrderingScheme> orderingScheme = Optional.empty();
-            if (!orderings.isEmpty()) {
-                orderingScheme = Optional.of(new OrderingScheme(orderBySymbols.build(), orderings));
+            if (!sortOrders.isEmpty()) {
+                OrderingScheme ordering = new OrderingScheme(orderBySymbols.build(), sortOrders);
+
+                if (partitionBySymbols.build().isEmpty()) {
+                    // if ORDER BY but no PARTITION BY, insert global sort step before
+                    subPlan = subPlan.withNewRoot(new SortNode(idAllocator.getNextId(), subPlan.getRoot(), ordering));
```

Comment:
Keep in mind that window operator still needs to collect all rows to compute some aggregates which require total partition count, so we don't really win on memory usage here. We win on performance though.

We've been thinking on adding rolling spill support for large window partitions. For example, we try to collect all rows for partition but when revoke is triggered we dump content which we re-read again when partition is finished. By that time we know partition count.

Reply:
> we don't really win on memory usage here [...]
> although we funnel all data through single node, singe CPU

I don't think single threadedness is avoidable here. But we certainly should not load everything into memory.

Before we add rolling spill, we should look into streaming window functions. For `row_number()` this is trivial (today, we have a dedicated node for that.) https://github.com/prestosql/presto/pull/464 attempts to cover the general case.

Reply:
> #464 attempts to cover the general case.

We should land that too. However, I don't think this removes necessity of having partition in memory. This is because for window aggregates number of preceding/following rows is also variable, so you don't know if some rows won't be needed later without doing full first pass on partition.

> Before we add rolling spill, we should look into streaming window functions.

This requires differentiations between window aggregates that require number of partition rows and which do not.

Reply:
When you insert explicit sort you should change `int preSortedChannelPrefix` parameter in `WindowNode` (it is set to 0 below). Otherwise sorting will happen anyway in `WindowOperator`

Reply:
`preSortedChannelPrefix` gets filled in at some point automatically (AddExchanges?)